File: ps/q10061.cpp
#include <iostream>
#include <queue>
#include <map>

typedef long long ll;
const int LEN = 40'000;

ll gcd(ll a, ll b) { return !b ? a : gcd(b, a % b); }
struct Fraction {
	ll num, den;
	Fraction(ll n, ll d) : num(n), den(d) { if (den < 0) num *= -1, den *= -1; if (!num) den = 1; }
	Fraction(const Fraction& f) : num(f.num), den(f.den) {}
	Fraction& operator=(const Fraction& f) { if (this != &f) num = f.num, den = f.den; return *this; }
	Fraction normalize() const { ll g = gcd(std::abs(num), den); return Fraction(num / g, den / g); }
	Fraction inv() const { return Fraction(den, num); }
	Fraction operator+(const Fraction& r) const {
		ll d = den * r.den / gcd(den, r.den);
		ll n = num * d / den + r.num * d / r.den;
		return Fraction(n, d).normalize();
	}
	Fraction operator-() const { return Fraction(-num, den); }
	Fraction operator-(const Fraction& r) const { return *this + (-r); }
	Fraction operator*(const Fraction& r) const { return Fraction(num * r.num, den * r.den).normalize(); }
	Fraction operator/(const Fraction& r) const { return *this * r.inv(); }
	friend std::ostream& operator<<(std::ostream& o, const Fraction& f) { return (o << f.num << '/' << f.den); }
};

struct Pos {
	Fraction x, y;

};

int p[101];
int find(int i) { return p[i] < 0 ? i : p[i] = find(p[i]); }
int join(int a, int b) {
	a = find(a), b = find(b);
	if (a == b) return 0;
	p[a] += p[b]; p[b] = a;
	return 1;
}

int main() {

}

File: BOJ/q10061.rs
use std::cmp::Ordering;
use std::fmt;
use std::ops::{Add, Sub, Mul, Div};

// 최대공약수를 구하는 함수
fn gcd(a: i128, b: i128) -> i128 {
    if b == 0 { a } else { gcd(b, a % b) }
}

#[derive(Clone, Copy, Debug)]
struct Fraction {
    num: i128,
    den: i128,
}

impl Fraction {
    fn from(n: i128) -> Self {
        Fraction { n, 1 }
    }

    // 생성자
    fn new(n: i128, d: i128) -> Self {
        let mut num = n;
        let mut den = d;
        if den < 0 {
            num = -num;
            den = -den;
        }
        if num == 0 {
            den = 1;
        }
        Fraction { num, den }.normalize()
    }

    // 분수를 정규화 (약분)
    fn normalize(self) -> Self {
        let g = gcd(self.num.abs(), self.den);
        Fraction { num: self.num / g, den: self.den / g }
    }

    // 역수
    fn inv(self) -> Self {
        Fraction::new(self.den, self.num)
    }
}

// Add trait을 구현하여 + 연산 지원
impl Add for Fraction {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        let d = self.den * other.den / gcd(self.den, other.den);
        let n = self.num * (d / self.den) + other.num * (d / other.den);
        Fraction::new(n, d)
    }
}

// Sub trait을 구현하여 - 연산 지원
impl Sub for Fraction {
    type Output = Self;

    fn sub(self, other: Self) -> Self {
        self + (-other)
    }
}

// Mul trait을 구현하여 * 연산 지원
impl Mul for Fraction {
    type Output = Self;

    fn mul(self, other: Self) -> Self {
        Fraction::new(self.num * other.num, self.den * other.den)
    }
}

// Div trait을 구현하여 / 연산 지원
impl Div for Fraction {
    type Output = Self;

    fn div(self, other: Self) -> Self {
        self * other.inv()
    }
}

// 음수 변환을 위한 Neg trait 구현
impl std::ops::Neg for Fraction {
    type Output = Self;

    fn neg(self) -> Self {
        Fraction::new(-self.num, self.den)
    }
}

// 표시를 위한 Display trait 구현
impl fmt::Display for Fraction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}/{}", self.num, self.den)
    }
}

fn main() {
    let a = Fraction::new(3, 4);
    let b = Fraction::new(5, 6);
    
    let sum = a + b;
    let diff = a - b;
    let prod = a * b;
    let div = a / b;

    println!("a + b = {}", sum);
    println!("a - b = {}", diff);
    println!("a * b = {}", prod);
    println!("a / b = {}", div);
}